notes.lua

```diff
@@ -0,0 +1,322 @@
+}
+
+function NoteManager:init()
+    local edit_mode = self.note ~= nil
+
+    self:addviews{
+        widgets.Window{
+            frame={w=35,h=20},
+            frame_inset={t=1},
+            autoarrange_subviews=false,
+            subviews={
+                widgets.HotkeyLabel {
+                    key='CUSTOM_ALT_N',
+                    label='Name',
+                    frame={t=0},
+                    on_activate=function() self.subviews.name:setFocus(true) end,
+                },
+                text_editor.TextEditor{
+                    view_id='name',
+                    focus_path='notes/name',
+                    frame={t=1,h=3},
+                    frame_style=gui.FRAME_INTERIOR,
+                    frame_style_b=nil,
+                    init_text=self.note and self.note.name or ''
+                },
+                widgets.HotkeyLabel {
+                    key='CUSTOM_ALT_C',
+                    label='Comment',
+                    frame={t=4},
+                    on_activate=function() self.subviews.comment:setFocus(true) end,
+                },
+                text_editor.TextEditor{
+                    view_id='comment',
+                    frame={t=5,b=3},
+                    focus_path='notes/comment',
+                    frame_style=gui.FRAME_INTERIOR,
+                    init_text=self.note and self.note.comment or ''
+                },
+                widgets.Panel{
+                    view_id='buttons',
+                    frame={b=0,h=3},
+                    autoarrange_subviews=true,
+                    subviews={
+                        edit_mode and widgets.TextButton{
+                            view_id='Save',
+                            frame={h=1},
+                            label='Save',
+                            key='CUSTOM_ALT_S',
+                            on_activate=function() self:saveNote() end,
+                            enabled=function() return #self.subviews.name:getText() > 0 end,
+                        } or widgets.TextButton{
+                            view_id='Create',
+                            frame={h=1},
+                            label='Create',
+                            key='CUSTOM_ALT_S',
+                            on_activate=function() self:createNote() end,
+                            enabled=function() return #self.subviews.name:getText() > 0 end,
+                        },
+                        edit_mode and widgets.TextButton{
+                            view_id='delete',
+                            frame={h=1},
+                            label='Delete',
+                            key='CUSTOM_ALT_D',
+                            on_activate=function() self:deleteNote() end,
+                        } or nil,
+                    }
+                }
+            },
+        },
+    }
+end
+
+function NoteManager:createNote()
+    local x, y, z = pos2xyz(df.global.cursor)
+    if x == nil then
```

Comment:
use `guidm.getCursorPos()` instead of accessing `df.global.cursor` directly. That will get you `nil` if the cursor is invalid. `df.global.cursor.x` is *never* `nil`, so this check will not work as written.

https://github.com/DFHack/dfhack/blob/develop/library/lua/gui/dwarfmode.lua#L40

Reply:
it is `nil` thanks to `pos2xyz` - but your solution is better.

Reply:
ah, I forgot that pos2xyz had that logic in it